packages/core/core/src/BundleGraph.js

```diff
@@ -400,13 +400,50 @@ export default class BundleGraph {
     }
   }
 
+  bundlesWithCycles: Set<NodeId> = new Set();
+
+  shouldExitEarly(assetNodeId: NodeId, bundleNodeId: NodeId): boolean {
+    // Detects whether a bundle has a cycle.
+    // If so, exit traversal early when removing the asset's subgraph.
+    // NOTE: this is not an ideal solution. See https://github.com/parcel-bundler/parcel/issues/6446
+    if (this.bundlesWithCycles.has(bundleNodeId)) {
+      return true;
+    }
+
+    let shouldExitEarly = false;
+
+    let entryNodes = new Set(this._graph.getNodeIdsConnectedFrom(bundleNodeId));
+
+    this._graph.traverse((nodeId, context, actions) => {
+      let node = nullthrows(this._graph.getNode(nodeId));
+      if (node.type !== 'asset') {
+        return;
+      }
+      if (entryNodes.has(nodeId) && assetNodeId !== nodeId) {
+        this.bundlesWithCycles.add(bundleNodeId);
+        shouldExitEarly = true;
+        actions.skipChildren();
+        return;
+      }
+    }, assetNodeId);
+
+    return shouldExitEarly;
+  }
+
   removeAssetGraphFromBundle(asset: Asset, bundle: Bundle) {
     let bundleNodeId = this._graph.getNodeIdByContentKey(bundle.id);
     let assetNodeId = this._graph.getNodeIdByContentKey(asset.id);
 
     // Remove all contains edges from the bundle to the nodes in the asset's
     // subgraph.
+    var shouldExitEarly = this.shouldExitEarly(assetNodeId, bundleNodeId);
+
     this._graph.traverse((nodeId, context, actions) => {
+      if (shouldExitEarly) {
```

Comment:
To my understanding, currently it looks like in the case that there are no cycles, currently you need to traverse that bundle's "assetgraph" twice ?

But, is there a way you could reduce the amount of traversals done here? Like maybe marking nodes for removal and then if it turns out to be cyclic, don't remove them, otherwise remove them ? Not sure if this would work/ have much impact, since deduplicate already takes so long, but maybe it could simplify logic so that `removeAssetgrpahfrombundle()` contains all the logic needed to do just that"